01-node-tutorial/answers/11-fs-async.js

```diff
@@ -0,0 +1,55 @@
+const { writeFile } = require("fs");
+
+const file = "./temporary/asyncOutput.txt";
+
+//callback hell
+
+// console.log("at start");
+
+// writeFile(file, "This is line 1\n", (err) => {
+//   console.log("at point 1");
+
+//   if (err) {
+//     console.log("This error happened at point 1: ", err);
+//     return;
+//   } else {
+//     writeFile(file, "This is line 2\n", { flag: "a" }, (err) => {
+//       console.log("at point 2");
+//       if (err) {
+//         console.log("This error happened at point 2: ", err);
+//         return;
+//       } else {
+//         writeFile(file, "This is line 3\n", { flag: "a" }, (err) => {
+//           console.log("at point 3");
+//           if (err) {
+//             console.log("This error happened at point 2: ", err);
+//             return;
+//           } else {
+//             console.log("Success");
+//           }
+//         });
+//       }
+//     });
+//   }
+// });
+
+// console.log("at end");
+
+//one more option
```

Comment:
### Required Change
[Race Condition]
When you ran this code it would've appeared to work, but unfortunately its functionally incorrect. You can't easily see the issue because you wrote out the exact same string on each line.
In the assignment, you needed all 3 lines of written to the file **in order**, but here because the `writeLinesToFile` function is async lines 51, 52, and 53 can actually write to the file in any order.
This is called a race condition - the final outcome relies upon your async work occurring in a specific order, but nothing ensures they actually run in that order. 

Try the following change to see this happen:
```js
const writeLinesToFile = (line) =>
    writeFile(file, `${line}\n`, { flag: "a" }, (err, result) => {
        console.log("Done");
        if (err) {
            console.log(err);
        }
        if (result) {
            console.log(result);
        }
    });

writeLinesToFile("This should be the first line.");
writeLinesToFile("This should be the second line.");
writeLinesToFile("This should be the third line.");
```

Here the output file should be:
```
This should be the first line.
This should be the second line.
This should be the third line.
```

But if you run it a number of times, you'll see that the lines aren't always in that order.

Here's the part of the original instructions that should be of some help in finding how to do this the "right" way: 
> Third, for each line you write, you need to do the write for the line that follows it within the callback - otherwise the operations won’t happen in order.

Reply:
Yes, it is clear to me now. I am going to stick with the required code from the assignment.